### PR TITLE
social login with google

### PR DESCRIPTION
**Pull request by phamthanhnhan14**

### CODE REVIEW

File: .idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (NamPNQ) [inline .idea/vcs.xml:1]: You must add `.idea` folder to `.gitignore`

File: app.py [hunk 2]
```diff
@@ -7,16 +7,76 @@
...
+REDIRECT_URI = '/oauth2callback'
+
+SECRET_KEY = 'staging sample key'
 
 app = Flask(__name__)
 app.config.from_object('app_config')
 if os.getenv('ENV_STAGE', 'dev') == 'test':
     app.config.from_object('tests.settings')
 
+app.secret_key = SECRET_KEY
```
Comment (NamPNQ) [inline app.py]: Storage `SECRET_KEY` in `app_config`

File: app.py [hunk 3]
```diff
@@ -7,16 +7,76 @@
...
+google = oauth.remote_app('google',
+                          base_url='https://www.google.com/accounts/',
+                          authorize_url='https://accounts.google.com/o/oauth2/auth',
+                          request_token_url=None,
+                          request_token_params={'scope': 'https://www.googleapis.com/auth/userinfo.email',
+                                                'response_type': 'code'},
+                          access_token_url='https://accounts.google.com/o/oauth2/token',
+                          access_token_method='POST',
+                          access_token_params={'grant_type': 'authorization_code'},
+                          consumer_key=GOOGLE_CLIENT_ID,
```
Comment (NamPNQ) [inline app.py]: Replace `GOOGLE_CLIENT_ID` to `app.config['GOOGLE_CLIENT_ID']`

File: app.py [hunk 4]
```diff
@@ -7,16 +7,76 @@
...
+                          consumer_secret=GOOGLE_CLIENT_SECRET)
```
Comment (NamPNQ) [inline app.py]: Replace `GOOGLE_CLIENT_SECRET` to `app.config['GOOGLE_CLIENT_SECRET']`

File: app.py [hunk 5]
```diff
@@ -7,16 +7,76 @@
 """
 
 import os
-from flask import Flask, render_template, request
+from flask import Flask, render_template, request, session, url_for, redirect
 from models import db, User
+from flask_oauth import OAuth
+
+GOOGLE_CLIENT_ID = '513593066071-r10fps7htf4fvuvunad08p065rsbe6h9.apps.googleusercontent.com'
+GOOGLE_CLIENT_SECRET = ' dtAIGo1CZkw8Ibuk1jlGpK5G'
```
Comment (NamPNQ) [inline app.py]: Move `GOOGLE_CLIENT_ID`, `GOOGLE_CLIENT_SECRET` to `app_config`

File: app.py [hunk 6]
```diff
@@ -7,16 +7,76 @@
...
+                          request_token_params={'scope': 'https://www.googleapis.com/auth/userinfo.email',
+                                                'response_type': 'code'},
+                          access_token_url='https://accounts.google.com/o/oauth2/token',
+                          access_token_method='POST',
+                          access_token_params={'grant_type': 'authorization_code'},
+                          consumer_key=GOOGLE_CLIENT_ID,
+                          consumer_secret=GOOGLE_CLIENT_SECRET)
+
+@app.route('/social_login')
+def social_login():
```
Comment (NamPNQ) [inline app.py:38]: No need

File: app.py [hunk 7]
```diff
@@ -7,16 +7,76 @@
...
+    except URLError, e:
+        if e.code == 401:
+            # Unauthorized - bad token
+            session.pop('access_token', None)
+            return redirect(url_for('google_login'))
+        return res.read()
+
+    return res.read()
+
+@app.route('/google_login')
```
Comment (NamPNQ) [inline app.py:61]: Change url to `/social/login/google`

File: app.py [hunk 8]
```diff
@@ -7,16 +7,76 @@
...
+def google_login():
+    callback=url_for('authorized', _external=True)
+    return google.authorize(callback=callback)
+
+@app.route(REDIRECT_URI)
```
Comment (NamPNQ) [inline app.py]: change url to `/social/google/authorized`

File: app.py [hunk 9]
```diff
@@ -7,16 +7,76 @@
...
+            session.pop('access_token', None)
+            return redirect(url_for('google_login'))
+        return res.read()
+
+    return res.read()
+
+@app.route('/google_login')
+def google_login():
+    callback=url_for('authorized', _external=True)
+    return google.authorize(callback=callback)
```
Comment (NamPNQ) [inline app.py]: return google.authorize(callback=url_for('google_authorized', _external=True))

File: app.py [hunk 10]
```diff
@@ -7,16 +7,76 @@
...
+
+@app.route(REDIRECT_URI)
+@google.authorized_handler
+def authorized(resp):
```
Comment (NamPNQ) [inline app.py:67]: Change name function to `google_authorized`

**Comment by phamthanhnhan14** [comment]
Review again plz.

**Comment by tuanquanghpvn** [comment]
Đợi review lâu qúa. Tự closed luôn ^^ cho có cảm giác được merge.

File: app.py [hunk 11]
```diff
@@ -7,16 +7,72 @@
...
+                          request_token_params={'scope': 'https://www.googleapis.com/auth/userinfo.email',
+                                                'response_type': 'code'},
+                          access_token_url='https://accounts.google.com/o/oauth2/token',
+                          access_token_method='POST',
+                          access_token_params={'grant_type': 'authorization_code'},
+                          consumer_key=app.config['GOOGLE_CLIENT_ID'],
+                          consumer_secret=app.config['GOOGLE_CLIENT_SECRET'])
+
+@app.route('/social/login/google')
+def social_login():
```
Comment (NamPNQ) [inline app.py:38]: Remove this function

File: app.py [hunk 12]
```diff
@@ -7,16 +7,72 @@
...
+from flask_oauth import OAuth
+
+REDIRECT_URI = '/oauth2callback'
 
 app = Flask(__name__)
 app.config.from_object('app_config')
 if os.getenv('ENV_STAGE', 'dev') == 'test':
     app.config.from_object('tests.settings')
 
+app.secret_key = app.config['SECRET_KEY']
```
Comment (NamPNQ) [inline app.py:21]: No need

File: app.py [hunk 13]
```diff
@@ -7,16 +7,72 @@
 """
 
 import os
-from flask import Flask, render_template, request
+from flask import Flask, render_template, request, session, url_for, redirect
 from models import db, User
+from flask_oauth import OAuth
+
+REDIRECT_URI = '/oauth2callback'
```
Comment (NamPNQ) [inline app.py:14]: No need

File: requirements.txt [hunk 14]
```diff
@@ -3,3 +3,4 @@ flask-redis
 flask-script
 factory-boy
 nose
+Flask-OAuth
```
Comment (NamPNQ) [inline requirements.txt:6]: Lower case is beautiful

**Comment by NamPNQ** [comment]
@phamthanhnhan14: Sorry for late response, please finish all comment :D

**Pull request closed without merging**